cli/azd/pkg/project/scaffold_gen.go

```diff
@@ -239,3 +258,70 @@ func mapHostUses(
 
 	return nil
 }
+
+func setParameter(spec *scaffold.InfraSpec, name string, value string, isSecret bool) {
+	for _, parameters := range spec.Parameters {
+		if parameters.Name == name { // handle existing parameter
+			if isSecret && !parameters.Secret {
+				// escalate the parameter to a secret
+				parameters.Secret = true
+			}
+
+			// prevent auto-generated parameters from being overwritten with different values
+			if valStr, ok := parameters.Value.(string); !ok || ok && valStr != value {
+				// if you are a maintainer and run into this error, consider using a different, unique name
+				panic(fmt.Sprintf(
+					"parameter collision: parameter %s already set to %s, cannot set to %s", name, parameters.Value, value))
+			}
+
+			return
+		}
+	}
+
+	spec.Parameters = append(spec.Parameters, scaffold.Parameter{
+		Name:   name,
+		Value:  value,
+		Type:   "string",
+		Secret: isSecret,
+	})
+}
+
+// evalEnvSubst evaluates a value that may contain multiple envsubst expressions, returning the evaluated value.
+//
+// The value may contain multiple expressions, e.g.:
+//   - "Hello, ${world:=okay}!"
+//   - "${CAT} and ${DOG}"
+//   - "${DB_HOST:='local'}:${DB_USERNAME:='okay'}"
+//
+// Each expression is evaluated and replaced with a reference to a parameter in the infra spec.
+// generateAsSecret is used to determine if the parameter generated should be marked as a secret.
+func evalEnvSubst(value string, generateAsSecret bool, infraSpec *scaffold.InfraSpec) string {
```

Comment:
FWIW - It took me a while to fully understand the semantics on how the output string is formed here.  From looking at the implementation sometimes the value is surrounded by single quotes, sometimes it isn't and sometimes the `${}`'s end up sticking around.

I think that this is done because we are trying to generate text which is a bicep string value, and this method is trying to generate an idomatic bicep string value (which explains why in the single reference case we just emit a variable reference (e.g. `foo`) instead of a something like `'${foo}'`, since the linters yell at you for that.

Maybe adding something in the name of the function or the comment about how it is trying to write the value as an idomatic bicep string literal?  That might also make it clearer was to why we call `scaffold.BicepName` on the value.

For a long while I was just thinking this was actually going to evaluate the string to a ground value instead of rewriting it as an interperolated string in a different format.

Reply:
Agreed 💯 on the confusing names and documents here.

I updated the name to `genBicepParamsFromEnvSubst` to better reflect what it's doing. Updated the comments to describe the Bicep outputs.